server/src/main/java/org/elasticsearch/action/support/PlainActionFuture.java

```diff
@@ -448,4 +449,18 @@ public static <T, E extends Exception> T get(CheckedConsumer<PlainActionFuture<T
         e.accept(fut);
         return fut.actionGet(timeout, unit);
     }
+
+    /**
+     * Same as {@code PlainActionFuture} but for use with {@link RefCounted} result types. Unlike {@code PlainActionFuture} this future
+     * acquires a reference to its result. This means that the result reference must be released by a call to {@link RefCounted#decRef()}
+     * on the result before it goes out of scope.
+     * @param <R> reference counted result type
+     */
+    public static class ForRefCounted<R extends RefCounted> extends PlainActionFuture<R> {
```

Comment:
We will probably need something similar for listenable future as well at some point. I didn't add any tests here yet before we're cool with the approach, I can though obviously it's trivial :)